File: FIXME.py
#Дан односвязный линейный список и указатель на голову списка P1. Необходимо вставить значение M перед каждым вторым
# элементом списка, и вывести ссылку на последний элемент полученного списка P2.
# При нечетном числе элементов исходного списка в конец списка вставлять не надо.

# FIXME излишние комментарии
# FIXME в названиие класса добавить Class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# FIXME в названиие класса добавить Class
class LinkedList:
    def __init__(self):
        self.head = None
    
    # FIXME отсутствует описание функции в docstring
    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return

        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    # FIXME непонятное название переменной m
    # FIXME отсутствует описание функции в docstring
    # FIXME неоптимальная логика функции
    def paste(self, m):
        if not self.head or not self.head.next:
            return

        current = self.head
        count = 1
        
        while current.next:  # Проверяем, что есть следующий узел
            if count % 2 != 0:  # Пропускаем первый элемент, вставляем перед вторым
                new_node = Node(m)
                new_node.next = current.next
                current.next = new_node
                current = new_node.next
                count += 2
            else:
                current = current.next
                count += 1

    # FIXME отсутствует описание функции в docstring
    def get_last_element(self):
        current = self.head
        if not current:
            return None
        while current.next:
            current = current.next
        return current

    def print_list(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next
        print()

# FIXME отсутствует описание docstring
# FIXME организация кода не стандартотизирована
linked_list = LinkedList()

# FIXME возможность ввести ненатуральное значение
num_elements = int(input("Введите количество элементов для добавления в список: "))
for i in range(num_elements):
    data = input(f"Введите элемент {i+1}: ")
    linked_list.append(data)

# FIXME непонятное название переменной m
m = input("Введите значение m для вставки: ")
linked_list.paste(m)  # Выполняем вставку значений

print("Измененный список:")
linked_list.print_list()

# Получаем указатель на последний элемент
last_node = linked_list.get_last_element()

# FIXME ненужное действие по выводу значения элемента
if last_node:
    print(f"Последний элемент списка имеет значение: {last_node.data}, Ссылка : {last_node}")
else:
    print("Список пуст")

File: FIXME.cpp
/*
Все динамические структуры данных реализовывать через классы. Не использовать STL.  
Для каждой динамической структуры должен быть предусмотрен
стандартный набор методов - добавления/удаления/вывода элементов. Во всех задачах обязательно
наличие дружественного интерфейса. Ввод данных с клавиатуры.

 Дан односвязный линейный список и указатель на голову списка P1. Необходимо
вставить значение M перед каждым вторым элементом списка, и вывести ссылку на последний
элемент полученного списка P2. При нечетном числе элементов исходного списка в конец
списка вставлять не надо.
*/
#include <iostream>

// FIXME: Запрещено использование `using namespace std;`
// FIXME: Следует использовать std:: префикс.
// FIXME: Запрещается использование tab в качестве пробела
using namespace std;

// FIXME: Отсутвуют комментарии, о назначении класса
class Node {
public:
    // FIXME: Поля класса объявлено как public, а не private.
    // FIXME: Поля данных класса должны иметь суффикс подчёркивания (data_, next_).
    int data;

    Node* next;

    // FIXME: Одноаргументный конструктор рекомендуется объявлять explicit.
    Node(int value) : data(value), next(nullptr) {}
};

// FIXME: Отсутвуют комментарии, о назначении класса
class LinkedList {
private:
    // FIXME: Имя поля класса должно заканчиваться подчёркиванием (head_).
    Node* head;

public:
    // FIXME: Так как класс управляет динамической памятью, 
    // FIXME: нужно явно запретить копирование и оператор присваивания.
    LinkedList() : head(nullptr) {}

    // FIXME: Неверное наименование метода
    void add(int value) {
        Node* newNode = new Node(value);
        if (head == nullptr) {
            head = newNode;
        }
        else {
            Node* current = head;
            while (current->next != nullptr) {
                current = current->next;
            }
            current->next = newNode;
        }
    }

    // Метод для вставки значения M перед каждым вторым элементом
    // FIXME: Неверное наименование метода
    // FIXME: Непонятная переменная M
    // FIXME: Неоптимальный метод
    void insertBeforeEverySecond(int M) {
        Node* current = head;
        int count = 1;  // Счетчик для отслеживания позиции

        while (current != nullptr && current->next != nullptr) {
            if (count % 2 == 1) {  // Если текущий элемент — первый в паре
                Node* newNode = new Node(M); // Создаем новый узел со значением M
                newNode->next = current->next; // Новый узел указывает на следующий элемент
                current->next = newNode; // Текущий узел указывает на новый узел
                current = newNode->next; // Переходим к следующему элементу
            }
            else {
                current = current->next;
            }
            count++;
        }
    }

    // FIXME: Неверное наименование метода
    void print() {
        Node* current = head;           // Начинаем с головы списка
        while (current != nullptr) {    // Пока не дойдем до конца списка
            cout << current->data << " "; // Выводим данные текущего узла
            current = current->next;    // Переходим к следующему узлу
        }
        cout << endl;         // Переход на новую строку после вывода
    }

    // Метод для получения указателя на последний элемент списка.
    // FIXME: Неверное наименование метода.
    // FIXME: Комментарий должен описывать назначение функции перед объявлением.
    Node* getLastNode() {
        Node* current = head;           // Начинаем с головы списка
        while (current != nullptr && current->next != nullptr) { // Идем до последнего узла
            current = current->next;
        }
        return current;                 // Возвращаем указатель на последний узел
    }

    // FIXME: Неверное наименование метода.
    // FIXME: Метод управления памятью не должен выполнять ввод/вывод.
    void clearList() {
        while (head != nullptr) {       // Пока список не пуст
            Node* temp = head;          // Сохраняем текущую голову
            head = head->next;          // Перемещаем голову на следующий узел
            delete temp;                // Удаляем узел
        }
        cout << "Список очищен." << endl;
    }

    friend void insertBeforeEverySecond(LinkedList& list, int M);
};

// FIXME: Лишнее действие, можно вызывать метод напрямую, а не через функцию
void insertBeforeEverySecond(LinkedList& list, int M) {
    list.insertBeforeEverySecond(M);
}

int main() {
    // FIXME: Вывод в консоль неккоректен
    LinkedList list;  // Создаем список
    // FIXME: Непонятные переменные n и M
    int n, value, M;

    // FIXME: Отсутствует проверка корректности ввода данных.
    cout << "Введите количество элементов в списке: ";
    cin >> n;

    // FIXME: Отсутствует проверка корректности ввода данных.
    cout << "Введите элементы списка: ";
    for (int i = 0; i < n; ++i) {
        cin >> value;               // Вводим значение элемента
        list.add(value);            // Добавляем элемент в список
    }

    // FIXME: Отсутствует проверка корректности ввода данных.
    cout << "Введите значение M для вставки: ";
    cin >> M;
    
    // Вставка значения M перед каждым вторым элементом
    insertBeforeEverySecond(list, M);

    cout << "Список после вставки: ";
    list.print();

    // Получение указателя на последний элемент списка
    Node* lastNode = list.getLastNode();

    // FIXME: Вывод адресов памяти не является дружественным интерфейсом.
    if (lastNode != nullptr) {
        cout << "Указатель на последний элемент списка: " << lastNode << endl;
        // FIXME: Действие нетребуемое заданием
        cout << "Значение последнего элемента: " << lastNode->data << endl;
    }
    else {
        cout << "Список пуст." << endl;
    }

    list.clearList();

    // FIXME: Действие нетребуемое заданием
    // Проверка, что список пуст
    cout << "Проверка списка после очистки: ";
    list.print();

    return 0;
}
